### PR TITLE
add documentation draft, version 2

### PR DESCRIPTION
**Pull request by wlammen**
I think it is time to create a doc folder in metamath-exe with a documentation draft based on the discussion in Issue #32 .  If this draft is accepted, I will slowly start documenting the sources along Doxygen style, beginning with the elementary functions first that I recognize as being from the technical infrastructure layer.

The focus will be on documentation.  Simple changes that can be easily reviewed may accompany this documentation if they do not break current code.  If at all done this will mostly be grouping related functionality, improving on variable/function names, decoupling.

I am working full time again, so expect a slow pace.  I am not a native English speaker so look out for typos and awkward grammar.

### CODE REVIEW

**Comment by digama0** [comment]
I'm not sure this document actually belongs in the `doc` folder. I would rather see this information in a sub-heading of the `CONTRIBUTING.md` file which explains how to contribute changes and best practices in doing so. IMO the `doc` folder should be focused on documenting the code, and not documenting the documentation process (meta-documentation). But I don't feel strongly about it.

**Comment by wlammen** [comment]
CONTRIBUTING.md is part of the metamath/set.mm project.  I wrote the Document.md file with the metamath/metamath-exe project in mind, and especially the Doxygen part is very specific to this application.  The same holds for issue #33, which this PR is based upon.

I understand you want the same documentation rules across all projects.  Which is a good idea. But in order to make progress, and not allowing too many side aspects blocking this issue forever, I favour adding a specific document now, and postpone a broader view of this topic in a different issue, maybe on metamath/set.mm.

To account for your idea the introductory note covers this aspect now.

File: doc/Documentation.md [hunk 1]
```diff
@@ -1,10 +1,14 @@
-# Documentation Rules in Metamath
+# Documentation Rules for metamath.exe
 
-This is a **draft**.  Here are some links on how to improve on documentation in general:
+This is a **draft**.   It was suggested to move parts or all of this to a
+different location, in particular to metamath/set.mm/CONTRIBUTING.md.
+
+Here are some links on how to improve on documentation in general:
```
Comment (benjub) [inline doc/Documentation.md:7]: "to improve documentation"
("to improve on" means "to do better than")

File: doc/Documentation.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    recommend using GitHub's markdown extensions (GFM) as a format style
+    embedded in plain text.  This rich text formatting is designed to be still
+    legible in simple editors.  Its format is in detail documented
+    [here](https://github.github.com/gfm), and is an extension to the
+    widespread [Markdown](https://commonmark.org/help/) format.
+    
+4. Source code documentation within sources should follow the
+    [Doxygen](https://www.doxygen.nl/index.html) style, so documentation of
+    variables and functions can be generated automatically.  Doxygen extracts
+    marked comment blocks from C sources and create HTML pages based upon them.
```
Comment (benjub) [inline doc/Documentation.md:44]: creates

Comment (wlammen) [reply doc/Documentation.md]: thanks, done

**Comment by digama0** [comment]
> CONTRIBUTING.md is part of the metamath/set.mm project.

No, I mean a CONTRIBUTING.md file for this repo, which explains how to contribute to this repo in particular. This is fairly conventional for github projects.

**Comment by wlammen** [comment]
Moved the text into a new file CONTRIBUTING.md

**Comment by jkingdon** [comment]
I like the idea of a CONTRIBUTING.md file (this is is only partly a matter of convention, it also gets shown in the github user interface as described at https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors  - see for example https://github.com/metamath/set.mm/contribute or https://github.com/metamath/metamath-exe/contribute , in addition to a few other places this link shows up).

As for the contents proposed here, I guess worth adding as a starting place? I'm a little bit worried about writing a lot of elaborate standards for documentation which doesn't exist yet (do we have doxygen yet? and any scripts or practices about generating documents from it?). But in general terms what is being suggested in this pull request seems fine.

**Comment by wlammen** [comment]
The use of doxygen is extremely easy.  I prepared a simple Doxyfile in the metamath directory on my notebook, and creation of a hyperlinked help is now a matter of a few seconds.  Even without any further preparation of the sources you can already browse all header files as a minimum support.  In addition a CONTRIBUTING.md entry automatically appears on the menu bar of the help.  This was one reason why I immediately followed the suggestion of Mario and embedded the documentation rules within some dummy text.

The way doxygen works allows building up a documentation piece by piece, and more than one person can work at the same time on it.  I suggest to simply give it a try.  If it is not accepted, it is a matter of a few deletions to revert to the current state.

This PR formally adds rules that allow a doxygen installation. Before I proceed further I need the impression this method is sufficiently supported by the community.  The planned next PR in this sequence will add a Doxyfile and modify .gitignore for proper help generation.

**Comment by jkingdon** [comment]
> The use of doxygen is extremely easy. I prepared a simple Doxyfile in the metamath directory on my notebook, and creation of a hyperlinked help is now a matter of a few seconds. Even without any further preparation of the sources you can already browse all header files as a minimum support. In addition a CONTRIBUTING.md entry automatically appears on the menu bar of the help. This was one reason why I immediately followed the suggestion of Mario and embedded the documentation rules within some dummy text.
> 
> The way doxygen works allows building up a documentation piece by piece, and more than one person can work at the same time on it. I suggest to simply give it a try. If it is not accepted, it is a matter of a few deletions to revert to the current state.
> 
> This PR formally adds rules that allow a doxygen installation. Before I proceed further I need the impression this method is sufficiently supported by the community. The planned next PR in this sequence will add a Doxyfile and modify .gitignore for proper help generation.

I certainly don't object to doxygen (I think I used it a few projects ago although my memory is not super clear on the point) and especially given your intention to bring in the `Doxyfile` and the like soon, I'd say we should go for it.

**Review by jkingdon — APPROVED** [review]

**Comment by benjub** [comment]
Maybe with the addition of these files at the root, it would be clearer to put the program itself in a /src/ folder ?

**Comment by digama0** [comment]
Definitely. (I think I put that one on the "code cleanup" list, but it should be there if not.)

**Comment by jkingdon** [comment]
> Maybe with the addition of these files at the root, it would be clearer to put the program itself in a /src/ folder ?

Part of me is a tiny bit sad that we have enough "extra files" that this is appealing, but sure, putting it in src makes sense.

**Review by jkingdon — APPROVED** [review]

**Pull request merged**